src/field/vecops.cxx

```diff
@@ -396,250 +396,216 @@ const Field3D V_dot_Grad(const Vector3D &v, const Field3D &f) {
 
-    result.z = VDDX(vcn.x, a.z, outloc) + VDDY(vcn.y, a.z, outloc) + VDDZ(vcn.z, a.z, outloc);
-    result.z -= vcn.x*(metric->G1_13*a.x + metric->G2_13*a.y + metric->G3_13*a.z);
-    result.z -= vcn.y*(metric->G1_23*a.x + metric->G2_23*a.y + metric->G3_23*a.z);
-    result.z -= vcn.z*(metric->G1_33*a.x + metric->G2_33*a.y + metric->G3_33*a.z);
+    result.z = VDDX(vcn.x, a.z) + VDDY(vcn.y, a.z) + VDDZ(vcn.z, a.z);
+    result.z -= vcn.x * (metric_x->G1_13 * a.x + metric_y->G2_13 * a.y + metric_z->G3_13 * a.z);
+    result.z -= vcn.y * (metric_x->G1_23 * a.x + metric_y->G2_23 * a.y + metric_z->G3_23 * a.z);
+    result.z -= vcn.z * (metric_x->G1_33 * a.x + metric_y->G2_33 * a.y + metric_z->G3_33 * a.z);
 
     result.covariant = true;
-  }else {
-    result.x = VDDX(vcn.x, a.x, outloc) + VDDY(vcn.y, a.x, outloc) + VDDZ(vcn.z, a.x, outloc);
-    result.x += vcn.x*(metric->G1_11*a.x + metric->G1_12*a.y + metric->G1_13*a.z);
-    result.x += vcn.y*(metric->G1_12*a.x + metric->G1_22*a.y + metric->G1_23*a.z);
-    result.x += vcn.z*(metric->G1_13*a.x + metric->G1_23*a.y + metric->G1_33*a.z);
-
-    result.y = VDDX(vcn.x, a.y, outloc) + VDDY(vcn.y, a.y, outloc) + VDDZ(vcn.z, a.y, outloc);
-    result.y += vcn.x*(metric->G2_11*a.x + metric->G2_12*a.y + metric->G2_13*a.z);
-    result.y += vcn.y*(metric->G2_12*a.x + metric->G2_22*a.y + metric->G2_23*a.z);
-    result.y += vcn.z*(metric->G2_13*a.x + metric->G2_23*a.y + metric->G2_33*a.z);
-
-    result.z = VDDX(vcn.x, a.z, outloc) + VDDY(vcn.y, a.z, outloc) + VDDZ(vcn.z, a.z, outloc);
-    result.z += vcn.x*(metric->G3_11*a.x + metric->G3_12*a.y + metric->G3_13*a.z);
-    result.z += vcn.y*(metric->G3_12*a.x + metric->G3_22*a.y + metric->G3_23*a.z);
-    result.z += vcn.z*(metric->G3_13*a.x + metric->G3_23*a.y + metric->G3_33*a.z);
+  } else {
+    result.x = VDDX(vcn.x, a.x) + VDDY(vcn.y, a.x) + VDDZ(vcn.z, a.x);
+    result.x += vcn.x * (metric_x->G1_11 * a.x + metric_y->G1_12 * a.y + metric_z->G1_13 * a.z);
+    result.x += vcn.y * (metric_x->G1_12 * a.x + metric_y->G1_22 * a.y + metric_z->G1_23 * a.z);
+    result.x += vcn.z * (metric_x->G1_13 * a.x + metric_y->G1_23 * a.y + metric_z->G1_33 * a.z);
+
+    result.y = VDDX(vcn.x, a.y) + VDDY(vcn.y, a.y) + VDDZ(vcn.z, a.y);
+    result.y += vcn.x * (metric_x->G2_11 * a.x + metric_y->G2_12 * a.y + metric_z->G2_13 * a.z);
+    result.y += vcn.y * (metric_x->G2_12 * a.x + metric_y->G2_22 * a.y + metric_z->G2_23 * a.z);
+    result.y += vcn.z * (metric_x->G2_13 * a.x + metric_y->G2_23 * a.y + metric_z->G2_33 * a.z);
+
+    result.z = VDDX(vcn.x, a.z) + VDDY(vcn.y, a.z) + VDDZ(vcn.z, a.z);
+    result.z += vcn.x * (metric_x->G3_11 * a.x + metric_y->G3_12 * a.y + metric_z->G3_13 * a.z);
+    result.z += vcn.y * (metric_x->G3_12 * a.x + metric_y->G3_22 * a.y + metric_z->G3_23 * a.z);
+    result.z += vcn.z * (metric_x->G3_13 * a.x + metric_y->G3_23 * a.y + metric_z->G3_33 * a.z);
 
     result.covariant = false;
   }
 
-  result.setLocation(outloc);
```

Comment:
I think we still need to set the location of the resultant vector otherwise it's exactly `CELL_CENTRE` but inputs might both be at `CELL_ZLOW` for example.